File: sem3/map/Lab/ToyLanguage/src/Repository/Repository.java
package Repository;

import Exceptions.MyExceptions;
import Model.PrgState;

import java.io.BufferedWriter;
import java.io.FileWriter;
import java.io.IOException;
import java.io.PrintWriter;

/**
 * Created by Andrei on 21.10.2017.
 */
public class Repository implements IRepository {
    private PrgState[] states;
    private String logFilePath;
    private int currentState;

    public Repository() {
        states = new PrgState[10];
        this.currentState =0;
    }

    public Repository(String logFilePath) {
        this.logFilePath = logFilePath;
        states = new PrgState[10];
        this.currentState =0;

    }

    public PrgState[] getStates() {
        return states;
    }

    public void setStates(PrgState[] states) {
        this.states = states;
    }

    public int getCurrentState() {
        return currentState;
    }

    public void setCurrentState(int currentState) {
        this.currentState = currentState;
    }

    public void add(PrgState state){
        states[currentState++] = state;
    }

    @Override
    public void logPrgStateExec() throws MyExceptions {
        PrintWriter logFile;
        try {
            logFile = new PrintWriter(new BufferedWriter(new FileWriter(logFilePath, true)));
        }catch(IOException e){
            throw new MyExceptions("Log file can not be opened");
        }
        logFile.append(states[currentState-1].toString());
        logFile.close();

    }

    @Override
    public PrgState getCrtPrg() {
        return states[currentState-1];

    }

    @Override
    public String toString() {
        String str = "";

        for(PrgState state : states)
        {
            str = str + state.toString();
        }
        str = str + "\n\n";
        return str;
    }
}

File: sem3/map/~craciunf/c-sharp-toy-language-interpreter/repo/Repository.cs
/* package repo; */
/* import models.*; */
/* import java.util.*; */
/* import java.io.*; */
/* import java.nio.file.*; */
using System;
using System.IO;
using System.Collections;
using System.Collections.Generic;

namespace ToyLanguageInterpreter {
    public class Repository : IRepository {
        private List <PrgState> states;
        private String logFilePath;

        public void prepareLogFilePath(String logFilePath) {
            if(logFilePath == "") {
                Console.Write("Log file: ");
                logFilePath = Console.ReadLine();
            }
            this.logFilePath = logFilePath;

        }

        public Repository(String logFilePath) {
            prepareLogFilePath(logFilePath);
            this.states = new List <PrgState>();
        }

        public Repository(String logFilePath, PrgState initialState) {
            prepareLogFilePath(logFilePath);
            this.states = new List <PrgState>();
            this.states.Add(initialState);
        }

        public PrgState getCurrentState() {
            // LinkedList implements getLast(), but this is
            // a more general List.
            return states[states.Count-1];
        }

        public void logPrgStateExec() {
            /*
            try {
                // TODO: first file open overwrites file.
                PrintWriter logFile = new PrintWriter(new BufferedWriter(new FileWriter(this.logFilePath, true)));

                logFile.println(this.getCurrentState().toString());
                logFile.close();
            } catch(Exception e) {
                    System.err.println(e);
                    throw e;
            }*/

            PrgState state = getCurrentState();

            using (StreamWriter logFile = new StreamWriter(logFilePath, true)) {
                try {
                    logFile.WriteLine("Execution Stack\n");
                    foreach (IStmt statement in state.getExeStack()) {
                        logFile.WriteLine("  " + statement + "\n");
                    }
                    logFile.WriteLine("\n");

                    logFile.WriteLine("Symbol Table\n");
                    foreach (KeyValuePair<string, int> entry in state.getSymTable()) {
                        logFile.WriteLine("  " + entry.Key + " --> " + entry.Value + "\n");
                    }
                    logFile.WriteLine("\n");

                    logFile.WriteLine("File Table\n");
                    foreach (KeyValuePair<int,MyFile> entry in state.getFileTable()) {
                        logFile.WriteLine("  " + entry.Key + " --> " + entry.Value + "\n");
                    }
                    logFile.WriteLine("\n");

                    logFile.WriteLine("Output\n");
                    foreach (int output in state.getStdout()) {
                        logFile.Write("  %d\n", output);
                    }
                    logFile.WriteLine("-------------------\n");
                }
                catch (Exception) {
                    Console.WriteLine("error: could not write to the given file");
                }
            }
        }
    }
}
